File: modules/signal-protocol/ios/Stores/PersistentPreKeyStore.swift
//
//  PersistentPreKeyStore.swift
//  Pods
//
//  TilliT Native — Signal Protocol bindings
//
import Foundation
import LibSignalClient

class PersistentPreKeyStore: PreKeyStore {
    // PreKeys are LOCAL keys used by ANY remote user, so use fixed prefix
    private let keyPrefix = "local"
    private let keychain = KeychainHelper.shared

    init(directoryURL: URL) {
        // directoryURL is ignored - preKeys are global/local, not per-session
    }

    func loadPreKey(id: UInt32, context: StoreContext) throws -> PreKeyRecord {
        let key = preKeyKey(for: id)
        guard let data = keychain.load(for: key) else {
            throw SignalError.invalidKey("No pre-key with id \(id)")
        }
        return try PreKeyRecord(bytes: [UInt8](data))
    }

    func storePreKey(_ record: PreKeyRecord, id: UInt32, context: StoreContext) throws {
        let key = preKeyKey(for: id)
        let data = Data(record.serialize())
        let saved = keychain.save(data: data, for: key)
        if !saved {
            throw SignalError.internalError("Failed to persist preKey to Keychain")
        }
    }

    func removePreKey(id: UInt32, context: StoreContext) throws {
        let key = preKeyKey(for: id)
        keychain.delete(for: key)
    }

    private func preKeyKey(for id: UInt32) -> String {
        return "preKey-\(keyPrefix)-\(id)"
    }
}

File: modules/signal-protocol/android/src/main/java/expo/modules/signalprotocol/stores/PersistentPreKeyStore.kt
package expo.modules.signalprotocol.stores

import android.content.Context
import org.signal.libsignal.protocol.InvalidKeyIdException
import org.signal.libsignal.protocol.state.PreKeyRecord
import org.signal.libsignal.protocol.state.PreKeyStore

class PersistentPreKeyStore(
    keyPrefix: String,  // ignored - preKeys are global/local, not per-session
    context: Context
) : PreKeyStore {
    // PreKeys are LOCAL keys used by ANY remote user, so use fixed prefix
    private val keyPrefix = "local"
    private val keystore = KeystoreHelper.getInstance(context)

    override fun loadPreKey(id: Int): PreKeyRecord {
        val key = preKeyKey(id)
        val data = keystore.load(key) ?: throw InvalidKeyIdException("No pre-key with id $id")
        return PreKeyRecord(data)
    }

    override fun storePreKey(id: Int, record: PreKeyRecord) {
        val key = preKeyKey(id)
        val saved = keystore.save(key, record.serialize())
        if (!saved) {
            throw RuntimeException("Failed to persist preKey to EncryptedSharedPreferences")
        }
    }

    override fun containsPreKey(id: Int): Boolean {
        val key = preKeyKey(id)
        return keystore.exists(key)
    }

    override fun removePreKey(id: Int) {
        val key = preKeyKey(id)
        keystore.delete(key)
    }

    private fun preKeyKey(id: Int): String {
        return "preKey-$keyPrefix-$id"
    }
}
